libraries/AP_Hott_Telem/AP_Hott_Telem.cpp

```diff
@@ -0,0 +1,358 @@
+        uint8_t gps_time_hs;          //#36 UTC time 0.01s units
+        int16_t vel_east;             //#37 velocity east mm/s
+        uint8_t horiz_acc;            //#39 horizontal accuracy dm
+        uint16_t msl_altitude;        //#40 mean sea level altitude
+        uint8_t gps_fix_char2;        //#42 GPS fix character2??. display, 'D' = DGPS, '2' = 2D, '3' = 3D, '-' = no fix
+        uint8_t version;       //#43  0: GPS Graupner #33600
+        uint8_t status1;
+        uint8_t status2;
+        uint8_t stop_byte = 0x7d;
+    } msg {};
+
+    AP_GPS &gps = AP::gps();
+    Location loc;
+
+    {
+        WITH_SEMAPHORE(gps.get_semaphore());
+        loc = gps.location();
+        msg.flight_direction = uint16_t(gps.ground_course() * 0.5 + 0.5);
+        msg.gps_speed_kmh = uint16_t(gps.ground_speed() * 3.6 + 0.5);
+        float sacc, hacc;
+        if (gps.speed_accuracy(sacc)) {
+            msg.speed_acc = sacc * 100 + 0.5;
+        }
+        if (gps.horizontal_accuracy(hacc)) {
+            msg.horiz_acc = hacc * 100 + 0.5;
+        }
+    }
+
+    float lat = loc.lat * 1.0e-7;
+    float lon = loc.lng * 1.0e-7;
+
+    uint16_t dm, sec;
+    GPS_to_DMS(lat, msg.pos_NS, dm, sec);
+    msg.pos_NS_dm = dm;
+    msg.pos_NS_sec = sec;
+
+    GPS_to_DMS(lon, msg.pos_EW, dm, sec);
+    msg.pos_EW_dm = dm;
+    msg.pos_EW_sec = sec;
+
+    AP_AHRS &ahrs = AP::ahrs();
+    Vector2f home_vec;
+    float posD = 0;
+    {
+        WITH_SEMAPHORE(ahrs.get_semaphore());
+        if (ahrs.get_relative_position_NE_home(home_vec)) {
+            msg.home_distance = home_vec.length();
+        }
+        ahrs.get_relative_position_D_home(posD);
+        Vector3f vel;
+        if (ahrs.get_velocity_NED(vel)) {
+            msg.climbrate = uint16_t(30000.5 + vel.z * -100);
+            msg.climbrate3s = 120 + vel.z * -3;
+            msg.vel_north = vel.x * 1000 + 0.5;
+            msg.vel_east = vel.y * 1000 + 0.5;
+        }
+    }
+
+    msg.altitude = uint16_t(500.5 - posD);
+
+    msg.gps_satelites = gps.num_sats();
+    switch (gps.status()) {
+    case AP_GPS::NO_GPS:
+    case AP_GPS::NO_FIX:
+        msg.gps_fix_char = '-';
+        break;
+    case AP_GPS::GPS_OK_FIX_2D:
+        msg.gps_fix_char = '2';
+        break;
+    default:
+        msg.gps_fix_char = '3';
+        break;
+    }
+    msg.gps_fix_char2 = msg.gps_fix_char;
+
+    msg.home_direction = degrees(atan2f(home_vec.y, home_vec.x)) * 0.5 + 0.5;
+
+    AP_RTC &rtc = AP::rtc();
+    {
+        WITH_SEMAPHORE(rtc.get_semaphore());
+        uint16_t ms;
+        rtc.get_system_clock_utc(msg.gps_time_h, msg.gps_time_m, msg.gps_time_s, ms);
+    }
+
+    msg.msl_altitude = loc.alt * 0.01 + 0.5;
+
+    send_packet((const uint8_t *)&msg, sizeof(msg));
+}
+
+/*
+  send a packet out
+ */
+void AP_Hott_Telem::send_packet(const uint8_t *b, uint8_t len)
+{
+    uint8_t crc = 0;
+    while (len) {
+        hal.scheduler->delay_microseconds(BYTE_DELAY_US);
```

Comment:
Should we be starting a timeout based on when `tx_pending` returns false instead?